### PR TITLE
SP updates for SM2

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

Allow wolfSSL to build with SP implementations of SM2.
Updates to SP implementation of other code.
# Testing

How did you test?

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: wolfcrypt/src/sp_sm2_arm32.c [hunk 1]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (JacobBarthelmeh) [inline wolfcrypt/src/sp_sm2_arm32.c:1]: please add license header to these dummy files

File: configure.ac [hunk 2]
```diff
@@ -7206,6 +7203,15 @@ do
     ENABLED_SP_ECC=yes
     ENABLED_SP_SAKKE_1024=yes
     ;;
+  smallsm2)
```
Comment (JacobBarthelmeh) [inline configure.ac:7206]: ran into an issue when trying to build with smallsm2 (maybe is in master too with just sp small?)

```
./configure --enable-all --enable-sp=smallsm2 --enable-sp-asm CC="clang -fsanitize=address" --enable-sm2 --enable-sm3 && make 
```

```
  CC       wolfcrypt/src/src_libwolfssl_la-ed25519.lo
wolfcrypt/src/eccsi.c:1437:45: error: unused parameter 'key' [-Werror,-Wunused-parameter]
static int eccsi_mulmod_point_add(EccsiKey* key, const mp_int* n,
                                            ^
wolfcrypt/src/eccsi.c:1437:64: error: unused parameter 'n' [-Werror,-Wunused-parameter]
static int eccsi_mulmod_point_add(EccsiKey* key, const mp_int* n,
                                                               ^
wolfcrypt/src/eccsi.c:1438:20: error: unused parameter 'point' [-Werror,-Wunused-parameter]
        ecc_point* point, ecc_point* a, ecc_point* res, mp_digit mp, int map)
                   ^
wolfcrypt/src/eccsi.c:1438:38: error: unused parameter 'a' [-Werror,-Wunused-parameter]
        ecc_point* point, ecc_point* a, ecc_point* res, mp_digit mp, int map)
                                     ^
wolfcrypt/src/eccsi.c:1438:52: error: unused parameter 'res' [-Werror,-Wunused-parameter]
        ecc_point* point, ecc_point* a, ecc_point* res, mp_digit mp, int map)
                                                   ^
wolfcrypt/src/eccsi.c:1438:66: error: unused parameter 'mp' [-Werror,-Wunused-parameter]
        ecc_point* point, ecc_point* a, ecc_point* res, mp_digit mp, int map)
                                                                 ^
wolfcrypt/src/eccsi.c:1438:74: error: unused parameter 'map' [-Werror,-Wunused-parameter]
        ecc_point* point, ecc_point* a, ecc_point* res, mp_digit mp, int map)
                                                                         ^
wolfcrypt/src/eccsi.c:1468:1: error: control reaches end of non-void function [-Werror,-Wreturn-type]
}
^
8 errors generated.
```

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]
Couple minor things. Very nice though!

Comparing `./configure --enable-all --enable-sp --enable-sp-asm --enable-sm2 --enable-sm3 && make`

This PR with updated wolfsm :
```
------------------------------------------------------------------------------
 wolfSSL version 5.6.3
------------------------------------------------------------------------------
Math: 	Multi-Precision: Wolf(SP) word-size=64 bits=4096 sp_int.c
	Single Precision: ecc 256 384 521 rsa/dh 2048 3072 4096 asm sp_x86_64.c
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
ECC   [      SM2P256V1]   256  key gen     80900 ops took 1.001 sec, avg 0.012 ms, 80855.681 ops/sec
ECDHE [      SM2P256V1]   256    agree     57000 ops took 1.001 sec, avg 0.018 ms, 56933.216 ops/sec
ECDSA [      SM2P256V1]   256     sign     56200 ops took 1.001 sec, avg 0.018 ms, 56125.185 ops/sec
ECDSA [      SM2P256V1]   256   verify     49800 ops took 1.001 sec, avg 0.020 ms, 49735.541 ops/sec
Benchmark complete
```

Vs current master:

```
------------------------------------------------------------------------------
 wolfSSL version 5.6.3
------------------------------------------------------------------------------
Math: 	Multi-Precision: Wolf(SP) word-size=64 bits=4096 sp_int.c
	Single Precision: ecc 256 384 521 rsa/dh 2048 3072 4096 asm sp_x86_64.c
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
ECC   [      SM2P256V1]   256  key gen     20300 ops took 1.004 sec, avg 0.049 ms, 20228.069 ops/sec
ECDHE [      SM2P256V1]   256    agree     20800 ops took 1.002 sec, avg 0.048 ms, 20753.558 ops/sec
ECDSA [      SM2P256V1]   256     sign     16600 ops took 1.002 sec, avg 0.060 ms, 16573.829 ops/sec
ECDSA [      SM2P256V1]   256   verify     15800 ops took 1.002 sec, avg 0.063 ms, 15770.492 ops/sec
Benchmark complete
```

File: wolfcrypt/src/sp_sm2_arm32.c (hunk 1 above)
Comment (SparkiDev) [reply wolfcrypt/src/sp_sm2_arm32.c:21]: Added

File: configure.ac (hunk 2 above)
Comment (SparkiDev) [reply configure.ac:7206]: Fixed

**Comment by JacobBarthelmeh** [comment]
In re-reviewing ./wolfcrypt/test/testwolfcrypt is failing for me:

```
cd wolfsm && ./install.sh && cd ..
cd wolfssl && ./configure --enable-all --enable-sp --enable-sp-asm --enable-sm2 --enable-sm3 && make

./wolfcrypt/test/testwolfcrypt
...
...
OPENSSL (PKEY1) passed!
OPENSSL (EVP Sign/Verify) passed!
SM2 Verify
ECC      test failed!
 error L=27600
 [fiducial line numbers: 7874 24284 35684 47863]
Exiting main with return code: -1

```

Env. details:

```
$ uname -a
Linux jak-virtual-machine 6.2.0-34-generic #34~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Sep  7 13:12:03 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

gcc version 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04) 

```

Trying with clang gives the same unit test failure error. Also could see the error reported when removing the --enable-sp-asm flag.

**Comment by SparkiDev** [comment]
The problem came from sm2.c where I was using the SP implementation even though it was a custom curve.
Fixed now.

**Comment by JacobBarthelmeh** [comment]
That solved the ./wolfcrypt/test/testwolfcrypt report for me. Am seeing a unit test failure though when running 'make check' still though (same configure).

```
$ ./scripts/unit.test -225
starting unit tests...
 Begin API Tests
   225: test_wc_ecc_sm2_sign_hash_ex                        :
ERROR - tests/api.c line 23777 failed with:
    expected: wc_ecc_sm2_sign_hash_ex(hash, sizeof(hash), rng, key, smallR, s) == (-3)
    result:   -2 != -3

 failed (  0.00027)

ERROR - tests/api.c line 66708 failed with:
    expected: Test failed

    result:   ret 0


FAILURES:
   225: test_wc_ecc_sm2_sign_hash_ex

 End API Tests
```

**Comment by SparkiDev** [comment]
Fixed

**Review by JacobBarthelmeh — APPROVED** [review]

**Pull request merged**